### ISSUE TITLE
Per-sample gradients

### ISSUE THREAD

**Issue by altosaar** [apache/incubator-mxnet#7987]
Mxnet should support per-sample gradients. 

This is needed to train graphical models (for example, multiple samples of gradients are needed to reduce variance).

Here is the desired behavior (it's a hacky way inspired by this [thread](https://github.com/tensorflow/tensorflow/issues/675) - please let me know if there's an easier way to do this!):

```
In [3]: from mxnet import gluon, nd, autograd
In [4]: params = gluon.ParameterDict()
In [5]: w = params.get('var', shape=(1,), grad_req='write')
In [6]: params.initialize()
In [7]: w.repeated = nd.repeat(w.data(), repeats=3, axis=0)
In [8]: w.repeated.attach_grad()
In [9]: with autograd.record():
   ...:     y = nd.random_uniform(shape=3) * w.repeated
   ...:     y.backward()
   ...:
In [10]: w.repeated.grad
Out[10]:
[ 0.60276335  0.85794562  0.54488319]
<NDArray 3 @cpu(0)>
```

Here is the current behavior:

```
In [11]: with autograd.record():
    ...:     y = nd.random_uniform(shape=3) * w.data()
    ...:     y.backward()
    ...:
In [12]: w.grad()
Out[12]:
[ 1.89447021]
<NDArray 1 @cpu(0)>
```

It is necessary for mxnet to support this feature - is this planned?

Thanks!

**Comment by piiswrong** [comment]
For now the best way to do this is to run forward/backward 3 times each time with 1 sample.

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by b0noI** [comment]
Suggested tags: "Feature request"
